vaadin-testbench-core/src/main/java/com/vaadin/testbench/unit/ComponentQuery.java

```diff
@@ -0,0 +1,212 @@
+     *            the id to look up
+     * @return the wrapper for the component with the given id
+     *
+     * @throws NoSuchElementException
+     *             if no component is found
+     */
+    public <X extends ComponentWrap<T>> X id(String id) {
+        Objects.requireNonNull(id, "id must not be null");
+        locatorSpec.id = id;
+        locatorSpec.count = new IntRange(1, 1);
+        return find();
+    }
+
+    /**
+     * Executes the search against current context and returns a list of test
+     * wrappers for matching components.
+     *
+     * @return a list of test wrappers for found components, or an empty list if
+     *         search does not produce results. Never {@literal null}.
+     */
+    @SuppressWarnings("unchecked")
+    public <X extends ComponentWrap<T>> List<X> all() {
+        return allComponents().stream()
+                .map(component -> (X) wrapperFactory.apply(component))
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Executes the search against current context and returns a list of
+     * matching components.
+     *
+     * @return a list of found components, or an empty list if search does not
+     *         produce results. Never {@literal null}.
+     */
+    public List<T> allComponents() {
+        if (context != null) {
+            return LocatorKt._find(context, componentType,
+                    spec -> Unit.INSTANCE);
+        }
+        return LocatorKt._find(componentType, spec -> Unit.INSTANCE);
+    }
+
+    /**
+     * Sets the context to search inside.
+     *
+     * If a {@literal null} value is given, the search will be performed againt
+     * the UI.
+     *
+     * @param context
+     *            a component used as starting element for search.
+     * @return this component query instance for chaining.
+     */
+    public ComponentQuery<T> from(Component context) {
+        this.context = context;
+        return this;
+    }
+
+    @SuppressWarnings("unchecked")
+    protected <X extends ComponentWrap<T>> X find() {
+        try {
+            return (X) wrapperFactory.apply(findComponent());
+        } catch (AssertionError e) {
+            throw new NoSuchElementException(e.getMessage());
+        }
+    }
+
+    protected T findComponent() {
+        if (context != null) {
+            LocatorKt._get(context, componentType, locatorSpec::populate);
+        }
+        return LocatorKt._get(componentType, locatorSpec::populate);
+    }
+
+    /**
+     * Private locator content holder used for searching the component.
+     *
+     * @param <T>
+     *            component type for search object.
+     */
+    private static class LocatorSpec<T extends Component> {
+
+        public String id;
+        public String caption;
+        public String placeholder;
+        public String text;
+        public IntRange count = new IntRange(0, Integer.MAX_VALUE);
+        public Object value;
+        public String classes;
```

Comment:
Should there be a way to search for "contains class" or "has theme"?

Reply:
Thanks for your feedback!
The implementation is still in progress. Additional query helpers will be added, but perhaps in other PRs.